packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,197 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/foundation.dart';
+
+import 'package:flutter/cupertino.dart';
+
+const EdgeInsets _kHorizontalItemPadding =
+    const EdgeInsets.symmetric(horizontal: 16.0, vertical: 0.0);
+
+const Radius defaultCornerRadius = const Radius.circular(3.0);
+
+/// An iOS-style segmented control bar
+///
+/// Displays the provided [options] widgets in a horizontal list.
+/// Exactly one [options] widget is always selected. By default,
+/// the leftmost widget is selected.
+///
+/// All widgets have the same width.
+///
+/// All widgets provided should have unique Key property.
+/// These keys will be used to identify the currently selected
+/// widget.
+///
+/// See also:
+///
+/// * <https://developer.apple.com/design/human-interface-guidelines/ios/controls/segmented-controls/>
+class SegmentedControl extends StatefulWidget {
+  /// Creates an iOS-style segmented control bar.
+  const SegmentedControl(
+      {Key key, @required this.options, @required this.initialValue})
+      : super(key: key);
+
+  /// The list of widgets of that will be used to populate the tabs of the bar.
+  final List<Widget> options;
+
+  /// The Key value of the widget that should be initially selected.
+  ///
+  /// This Key should be the Key of one of the widgets within [options].
+  final Key initialValue;
+
+  @override
+  _SegmentedControlState createState() => new _SegmentedControlState();
+}
+
+class _SegmentedControlState extends State<SegmentedControl> {
+  Key sharedState;
+
+  @override
+  void initState() {
+    super.initState();
+    sharedState = widget.initialValue;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    List<Widget> buttons = new List<Widget>();
+
+    for (int i = 0; i < widget.options.length; i += 1) {
+      buttons.add(new Expanded(
+          child: new _SegmentedControlButton(
+              widget.options.elementAt(i), widget.options.elementAt(i).key,
+              (Key state) {
+        setState(() {
```

Comment:
Ahh, I see.  You shouldn't need to call setState: the widget that embeds things would do that on itself, in response to you calling their valueChanged callback.  You just pass their callback in here. When they call their setState, this widget will get rebuilt with the new values.